syntax= "proto3";

service Market{
  rpc registerSeller(registerSellerReq) returns (registerSellerRes){};
  rpc sellItem(sellItemReq) returns (sellItemRes){};
  rpc addProduct(Product) returns (void){};
  rpc updateProduct(UpdateItemReq) returns (UpdateItemRes){};
  rpc deleteProduct(DeleteItemReq) returns (DeleteItemRes){};
  rpc viewProduct(void) returns (Product){};
  rpc searchProduct(SearchReq) returns (Products){};
  rpc addRating(Rate_an_item) returns (UpdateItemRes){};
  rpc displayProducts(Seller) returns (Products){};
  rpc buyProduct(BuyItemReq) returns (BuyItemRes){};
  rpc addtoWishlist(WishListReq) returns (WishListRes){};
  rpc NotifyBuyer(NotificationReq) returns (NotificationRes){};
  rpc NotifySeller(NotificationReq) returns (NotificationRes){};
  

}

message void{
}

enum Status {
  SUCCESS = 0;
  FAILURE = 1;
}

enum Category {
  ELECTRONIC=0;
  FASHION=1;
  OTHERS=2;
}

message Buyer{
  string UUID = 1;
  string address = 2;
  repeated Product wishlist=3;
  string notif_ip=4;
}

message Seller{
  string UUID = 1;
  string address = 2;
  repeated Product products=3;
  string notif_ip=4;
}

message Sellers{
  repeated Seller sellers = 1;
}

message Rate_an_item{
  string ProductUUID=1;
  string address=2;
  int32 rating=3;

}

message Product{
  string name=1;
  float price=2;
  int32 quantity=3;
  string description=4;
  string seller_address=5;
  string seller_UUID=6;
  string Product_UUID=7;
  string category =8;
  Seller seller=9;
  float rating=10;
  int32 rating_count=11;
  repeated Buyer wishlist = 12;

}

message Products{
  repeated Product products = 1;
}

message registerSellerReq{
  string address = 1;
  string uuid = 2;
  string notif_ip=3;

}

message registerSellerRes{
 Status status = 1;
}


message sellItemReq{
  string name = 1;
  int32 quantity = 2;
  string description = 3;
  string sellerAddress = 4;
  float price = 5;
  string sellerUUID = 6;
  string Category = 7;
  Seller seller=8;
}

message sellItemRes{
  string productUUID = 1;
  string status = 2;
  Seller seller=3;
}

message UpdateItemReq{
  string Product_UUID=1;
  string seller_UUID=2;
  string seller_address=3;
  int32 price=4;
  int32 quantity=5;
  string description=6;
  Seller seller=7;
}
message UpdateItemRes{
  string productUUID = 1;
  string status = 2;
}

message DeleteItemReq{
  string Product_UUID=1;
  string seller_UUID=2;
  string seller_address=3;

}

message DeleteItemRes{
  string productUUID = 1;
  string status = 2;

}
message WishListReq{
  string productUUID=1;
  string address=2;
  string notif_ip=3;
  string uuid=4;
}

message WishListRes{
  string productUUID = 1;
  string status = 2;

}

message BuyItemReq{
  string item_id=1;
  int32 quantity=2;
  string address=3;
}
message BuyItemRes{
  string productUUID = 1;
  string status = 2;
}

message SearchReq{
  string item_name=1;
  string category=2;
}
message NotificationReq{
  string address=1;
  string UUID=2;
  string notif_ip=3;

}

message NotificationRes{
  string message=1;
}

